Modifies validator to flag unusual values for the DATS 'formats' field

cmadjar opened the pull request:
## Description

In order to harmonize DATS between datasets, a review of the available content for the distributions.formats field has been done and the following convention were determined:
- file format should be capitalized without the dot preceding the format (exceptions being BigWig, NIfTI and RNA-Seq)

Examples of invalid format: `tsv`, `.txt`...
Examples of a valid format: TSV, TXT, MINC...


## Related issues (optional)

https://github.com/CONP-PCNO/conp-dataset/issues/491

cmadjar commented:
In theory, all datasets should pass the format validation except "Learning_Naturalistic...".

cmadjar commented:
@zxenia @emmetaobrien if you could review, that would be awesome! I tested locally and it seems to work fine. 

Thank you!

cmadjar commented:
@emmetaobrien I don't understand. It keeps telling me that 1000Genomes is not correct and has a dot in front of VCF but if I look at the DATS.json, the dot is not there... Am I missing something? I'm so confused

@zxenia it looks like the check for formats is skipped for crawled datasets, any idea why and how I could fix this?

Thank you both!

emmetaobrien commented:
That issue with 1000GenomesProject does not make sense to me, the file appears correctly formatted and I have just run the freestanding validator on it and it comes back entirely correct.

zxenia commented:
@cmadjar @emmetaobrien 
I run the validator with your changes on two files: valid_dats.json (which has 1000Genomes metadata inside) in the examples folder and the DATS.json from the 1000Genomes folder in /projects dir.
In both cases, it reports that the files are invalid because of ".VCF" in the subdatasets that are in "hasPart" section. Indeed the formats there have ".VCF" value. In dats validator we validate recursively all nested datasets because they must follow DATS schema too. So it means that everything works as expected.

zxenia commented on scripts/dats_validator/examples/valid_dats.json:
```diff
@@ -68,7 +68,7 @@
     {
       "formats": [
-        ".VCF",
+        "VCF",
```
@cmadjar You have to change ".VCF" to "VCF" everywhere (subdatasets in hasPart)  in the file to make the file valid.

zxenia commented on scripts/dats_validator/validator.py:
```diff
@@ -105,11 +106,42 @@ def validate_extra_properties(dataset):
+def validate_formats(dataset):
+    """ Checks if the values in the formats field of the JSON object follows the upper case convention without dots. """
+
+    errors_list       = []
```
multiple spaces before operator =
it  should be one space

zxenia reviewed (changes requested):
Currently, CircleCI TEST failing in `test_dats_validator` because the `valid_dats.json` was changed and it is not valid now. Once it's fixed the test should pass.

zxenia commented:
> it looks like the check for formats is skipped for crawled datasets, any idea why and how I could fix this?

@cmadjar Could please you point me to where/or which crawled datasets are being skipped?

cmadjar commented:
@zxenia Thank you for the fix suggestion! I updated the example DATS and modified their titles to avoid confusions in the future.

Regarding the dataset not being flag, here is the one I know should be failing the tests: https://github.com/conp-bot/conp-dataset-Learning_Naturalistic_Structure__Processed_fMRI_dataset/tree/6125592fbcaf1abfe7ab6ee9cfcb93ee587a3c7c

zxenia commented on scripts/dats_validator/validator.py:
```diff
@@ -105,11 +106,54 @@ def validate_extra_properties(dataset):
+                            f"Please add a the 'formats' property to 'distributions'."
+            errors_list.append(error_message)
+
+    if errors_list:
+        return False, errors_list
```
@cmadjar The function returns `False` and `errors_list` in two places, this is redundant. You can rearrange the if condition. First, you check if 'formats' is not there, then you can add the rest in else. 

for example:

```
format_exceptions = ['BigWig', 'NIfTI', 'RNA-Seq']
for distribution_dict in dataset['distributions']:
    if 'formats' not in distribution_dict:
        error_message = f"Validation error in {dataset['title']}: distributions." \
                                    f"formats - 'formats' property is missing under distributions. " \
                                    f"Please add a the 'formats' property to 'distributions'."
        errors_list.append(error_message)
    else:
        for file_format in distribution_dict['formats']:
            if file_format != file_format.upper() and file_format not in format_exceptions:

            ...... etc. ........

if errors_list:
    return False, errors_list
else:
    return True, errors_list
```

zxenia commented on tests/template.py:
```diff
@@ -10,6 +10,7 @@
 import pytest
 
 from scripts.dats_validator.validator import validate_json
+from scripts.dats_validator.validator import validate_non_schema_required
```
add import on the line above after  comma

cmadjar commented:
@zxenia thank you! Your comments should be included in the last commit.

cmadjar commented:
The changes to the validator flags the following errors that should be resolved:
- [ ] bigbrain-datalad: extra property subjects is missing
  - [x] https://github.com/conpdatasets/bigbrain-datalad/pull/1
- [ ] BigBrain_3DSurfaces: extra property subjects is missing
- [ ] BigBrain 3DClassifiedVolumes: missing subjects
- [ ] Khanlab/BigBrainMRICoreg: extra property subjects is missing
- [ ]  Khanlab/BigBrainHippoUnfold: missing subjects
- [ ] Mulit-model functionalization of disease associated PTEN missense mutations: subjects is missing
- [ ] Calgary Campinas: formats error (nii.gz, .h5, .zip)
- [ ] mousebytes: formats error (csv)
- [ ] nhpatlas: subjects missing * 9
- [ ] openpain cbp resting: subjects missing
- [ ] openpain placebo tetrault: subjects missing
- [ ] openpain thermal: subjects missing
- [ ] openpain_BrainNetworkChange_Mano: subjects missing
- [ ] openpain_placebo_1: subjects missing
- [ ] refseq Callitrhix: subjects missing
- [ ] refseq Mus: subjects missing
- [ ] refseq Homo: subjects and files missing
- [ ] Multimodal data with wide field GCaMP imaging: missing subjects, CONP status and formats
- [ ] Reusing Neuro Data: missing subjects
- [ ] Systematic phenomics analysis of autism-associated genes: subjects missing
- [ ] celltypes: formats error: csv
- [ ] mcgill-emc-rna-seq-experiment: formats error: bigWig mislabelled



Crawlers datasets:
- [ ] CFMM 7T MP2RAGE T1 mapping: formats, subjects and CONP status
- [ ] Calgary Preschool MRI dataset: formats, subjects and CONP status
- [ ] Intracellular Recordings of Murine Neocortical Neurons: formats, subjects and CONP status
- [ ] Learning Naturalistic structure processed fMRI dataset: formats not following convention, subjects missing
- [ ] Numerically Perturbed Structural Connectomes from ...: subjects and CONP status missing
- [ ] Quantifying-Neural-Cognitive-Relationships-Across-the-Brain: formats, subjects and CONP status
- [ ] VFA T1 mapping...: subjects and CONP status
- [ ] Comparing Perturbation Modes for evaluating... subjects is missing
- [ ] MRI and unbiased averages of ....: subjects and CONP status is missing

emmetaobrien commented:
`mcgill-emc-rna-seq-experiment` is not in error; only the W in "bigWig" should be capitalised.

zxenia commented on scripts/dats_validator/validator.py:
```diff
@@ -105,11 +106,50 @@ def validate_extra_properties(dataset):
+                            f"Please add the 'formats' property to 'distributions'."
+            errors_list.append(error_message)
+        else:
+            file_formats_list = reduce(
+                lambda x,y: x+y,
+                [ prop['formats'] for prop in dataset['distributions'] ]
+            )
+            for file_format in file_formats_list:
```
Now there is no need for reduce() here, you can access each distribution_dict object and its 'formats' list directly. reduce() returns a list of all formats across all distribution_dicts in one dataset, but we are in a `for` loop for each distribution object. Using reduce() will throw KeyError again if 'formats' is not present in one of them. All that's needed here is:
`else: for file_format in distribution_dict['formats']:` instead of the selected lines.

cmadjar replied:
right. Should be resolved in the last commit. I will also need to modify the validation of extraProperties for crawled data as subjects and CONP status cannot be automatically populated by the crawlers... So we need to skip the validation of those two required fields.
Will do that tomorrow

emmetaobrien commented:
I have corrected the format issues in the DATS.json for `celltypes` and the two `Khanlab/BigBrain*` datasets, and contacted the appropriate person to inquire about the number of subjects for the the `openpain` datasets.

emmetaobrien commented:
I have now also corrected DATS.json in `mousebytes` and `calgary-campinas`.

emmetaobrien commented:
Have just updated DATS.json for the three `refseq_*` datasets, with a value of 0 for subjects as a subject count for a reference genome does not really make sense.

cmadjar commented:
pulled by mistake some dataset on that branch and I am not sure how to revert cleanly the pulled dataset so I will be closing this PR in favour of https://github.com/CONP-PCNO/conp-dataset/pull/507 that contains only the script changes.
